modules/artifact-activation/src/main/java/org/mule/runtime/module/artifact/activation/api/ast/ArtifactAstUtils.java

```diff
@@ -107,43 +147,106 @@ public static ArtifactAst parseAndBuildAppExtensionModel(String[] configResource
-              .addParameter(MULE_SDK_TYPE_LOADER_PROPERTY_NAME, new ApplicationTypeLoader(dependenciesExtensionModels))
+              .addParameter(MULE_SDK_EXTENSION_NAME_PROPERTY_NAME, artifactId)
               .build()));
     } else {
-      logModelNotGenerated("Mule ExtensionModelLoader not found", muleContext);
+      logModelNotGenerated("Mule ExtensionModelLoader not found", artifactId, ast.getArtifactType());
       return empty();
     }
   }
 
+  /**
+   * Handles a {@link ValidationResult} uniformly. Warnings are logged, errors raise exceptions.
+   *
+   * @param validationResult The {@link ValidationResult} obtained using an {@link ArtifactAstValidator}.
+   * @throws ConfigurationException If there was any result item with {@link ERROR} level.
+   */
+  public static void handleValidationResult(ValidationResult validationResult)
+      throws ConfigurationException {
+    final Collection<ValidationResultItem> items = validationResult.getItems();
+
+    // Logs warnings if any.
+    items.stream()
+        .filter(v -> v.getValidation().getLevel().equals(WARN))
+        .forEach(v -> LOGGER.warn(validationResultItemToString(v)));
+
+    // If there are any errors, throws.
+    final boolean hasErrors = items.stream().anyMatch(v -> v.getValidation().getLevel().equals(ERROR));
+    if (hasErrors) {
+      throw new ConfigurationException(createStaticMessage(validationResult.getItems()
+          .stream()
+          .map(ArtifactAstUtils::validationResultItemToString)
+          .collect(joining(lineSeparator()))));
+    }
+  }
+
+  private static String validationResultItemToString(ValidationResultItem v) {
+    return v.getComponents().stream()
+        .map(component -> component.getMetadata().getFileName().orElse("unknown") + ":"
+            + component.getMetadata().getStartLine().orElse(-1))
+        .collect(joining("; ", "[", "]")) + ": " + v.getMessage();
+  }
+
   private static ArtifactAst doParseArtifactIntoAst(String[] configResources,
                                                     AstXmlParserSupplier parserSupplier,
                                                     Set<ExtensionModel> extensions,
-                                                    boolean disableValidations) {
-    try {
-      return parserSupplier.getParser(extensions, disableValidations).parse(loadConfigResources(configResources));
-    } catch (Exception e) {
-      throw new MuleRuntimeException(e);
+                                                    boolean disableValidations)
+      throws ConfigurationException {
+    return parserSupplier.getParser(extensions, disableValidations).parse(loadConfigResources(configResources));
+  }
+
+  private static boolean definesOwnExtensionModel(ArtifactAst artifactAst) {
+    ArtifactType artifactType = artifactAst.getArtifactType();
+    switch (artifactType) {
+      case APPLICATION:
+        return artifactAst.topLevelComponentsStream()
```

Comment:
this is conceptually wrong. If the application doesn't define any reusable components it doesn't mean it doesn't have an extension model. It just means it has one with no components. There's still value in having an extension model that simply exposes types, or to at least being able to assume that the extension model will always be there.

Not a fan of the method name either.